src/pam/main.cc

```diff
@@ -0,0 +1,394 @@
+    return PAM_SYSTEM_ERR;
+  }
+
+  // Stop executing if Howdy has been disabled in the config
+  if (reader.GetBoolean("core", "disabled", false)) {
+    syslog(LOG_INFO, "Skipped authentication, Howdy is disabled");
+    return PAM_AUTHINFO_UNAVAIL;
+  }
+
+  // Stop if we're in a remote shell and configured to exit
+  if (reader.GetBoolean("core", "ignore_ssh", true)) {
+    if (getenv("SSH_CONNECTION") != nullptr ||
+        getenv("SSH_CLIENT") != nullptr || getenv("SSHD_OPTS") != nullptr) {
+      syslog(LOG_INFO, "Skipped authentication, SSH session detected");
+      return PAM_AUTHINFO_UNAVAIL;
+    }
+  }
+
+  // Try to detect the laptop lid state and stop if it's closed
+  if (reader.GetBoolean("core", "ignore_closed_lid", true)) {
+    glob_t glob_result{};
+
+    // Get any files containing lid state
+    int return_value =
+        glob("/proc/acpi/button/lid/*/state", 0, nullptr, &glob_result);
+
+    // TODO: We ignore the result
+    if (return_value != 0) {
+      globfree(&glob_result);
+    }
+
+    for (size_t i = 0; i < glob_result.gl_pathc; i++) {
+      ifstream file(string(glob_result.gl_pathv[i]));
+      string lid_state;
+      getline(file, lid_state, (char)file.eof());
+
+      if (lid_state.find("closed") != std::string::npos) {
+        globfree(&glob_result);
+
+        syslog(LOG_INFO, "Skipped authentication, closed lid detected");
+        return PAM_AUTHINFO_UNAVAIL;
+      }
+    }
+
+    globfree(&glob_result);
+  }
+
+  // If enabled, send a notice to the user that facial login is being attempted
+  if (reader.GetBoolean("core", "detection_notice", false)) {
+    if ((pam_res = conv_function(
+             PAM_TEXT_INFO,
+             dgettext("pam", "Attempting facial authentication"))) !=
+        PAM_SUCCESS) {
+      syslog(LOG_ERR, "Failed to send detection notice");
+    }
+  }
+
+  // Get the username from PAM, needed to match correct face model
+  char *user_ptr = nullptr;
+  if ((pam_res = pam_get_user(pamh, (const char **)&user_ptr, nullptr)) !=
+      PAM_SUCCESS) {
+    syslog(LOG_ERR, "Failed to get username");
+    return pam_res;
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  posix_spawn_file_actions_init(&file_actions);
+  // We close standard descriptors for the child
+  posix_spawn_file_actions_addclose(&file_actions, STDOUT_FILENO);
+  posix_spawn_file_actions_addclose(&file_actions, STDERR_FILENO);
+  posix_spawn_file_actions_addclose(&file_actions, STDIN_FILENO);
+
+  const char *const args[] = {
+      "/usr/bin/python3", "/lib/security/howdy/compare.py", user_ptr, nullptr};
+  pid_t child_pid;
+
+  // Start the python subprocess
+  if (posix_spawnp(&child_pid, "/usr/bin/python3", &file_actions, nullptr,
+                   (char *const *)args, nullptr) < 0) {
+    syslog(LOG_ERR, "Can't spawn the howdy process: %s", strerror(errno));
+    return PAM_SYSTEM_ERR;
+  }
+
+  mutex m;
+  condition_variable cv;
+  Type confirmation_type;
+  // TODO: Find a clean way to do this
```

Comment:
Seems like a good usecase for [`std::atomic`](https://en.cppreference.com/w/cpp/atomic/atomic) [`Compare and Set`](https://en.wikipedia.org/wiki/Compare-and-swap).

Reply:
Thanks for the suggestion!